### ISSUE TITLE
Additional file editor syntax/languages?

### ISSUE THREAD

**Issue by MostHated** [webmin/webmin#1440]
Hello,
    I was wondering if there is anywhere to get additional language/syntax plugins for the text editor in the file manager? I am not sure if it uses some sort of standardized format in which I could grab a language from another application or possibly adapt it and add it to Webmin?

Thanks,
-MH

**Comment by iliarostovtsev** [comment]
Hi,

What language syntax highlight is missing?

**Comment by swelljoe** [comment]
It uses CodeMirror (https://codemirror.net/), though we have a five year old version of it (@iliaross @jcameron would it be difficult to update?).

Here's the languages the current version supports, but I dunno how many of those our old version supports or how to add to it: https://codemirror.net/mode/index.html

**Comment by iliarostovtsev** [comment]
Joe, Code mirror we distribute is pretty recent - few months old at most. What makes you think it's five years old?

**Comment by swelljoe** [comment]
https://github.com/webmin/webmin/tree/master/filemin/unauthenticated/js/lib/codemirror

**Comment by iliarostovtsev** [comment]
We don't use this with Authentic Theme.

**Comment by MostHated** [comment]
I appreciate the replies.

 I have been using the [Caddy web server](https://caddyserver.com) a lot lately, so support for their Caddyfile was what I was mainly after, similar to how there was an Nginx specific config setting.

I find it quite convenient to be able to open two side by side in webmin instead of having to open two separate terminal text editors, etc. Having the hotkey support for comment/block comment would be extremely helpful at times when trying things out.

**Comment by swelljoe** [comment]
So, two copies of codemirror seems sub-optimal? Can Authentic not just use the one in Webmin (and update the one in Webmin, so it's not ancient)?

**Comment by iliarostovtsev** [comment]
Perhaps it's sub-optimal but those custom bundles are most likely not the same. Besides, theme can use CodeMirror without Filemin module enabled.

Yes, I could update it for the old theme.

@MostHated Try she'll highlight - does it work for your case?

**Comment by iliarostovtsev** [comment]
*shell*. *

**Comment by MostHated** [comment]
I didn't think to try that one. It seems to work pretty well actually. 👍 

![](https://i.imgur.com/k0CZ1F3.png)

Highlighting seems to be decently functional, but most importantly ctrl + / works for single line as well as multiple lines. So that seems like it should do the trick. I definitely appreciate it.

**Comment by iliarostovtsev** [comment]
Yes, I know this because you mentioned Nginx and this is what we do for Nginx configs highlight. For Apache it's even more complex as it is a custom mod to highlight specific directives only for Apache (as it's very popular).

I will add specific config highlights (Apache, Nginx and etc) for pop-up file editor as well (for the next theme release). But it's pretty hard to do, as those config files are indistinguishable based on extension.

Tell me, what is the extension for those Caddy config files?

**Comment by MostHated** [comment]
They typically don't have one either,  unfortunately. But they are almost always named "Caddyfile" unless you specifically rename and pass the new file in.

**Comment by iliarostovtsev** [comment]
Okay, that's helpful. No extension by default. Where are those config files typically located?

**Comment by MostHated** [comment]
Well, it doesn't really have to live anywhere specific I don't believe. I typically run it via docker-compose, in which case the Caddyfile just sits along side my docker-compose.yml, or else I create a data folder and place it in there if I am feeling wild. If Caddy is actually installed on a system you typically just create a folder somewhere and create a Caddyfile in said folder. When you run Caddy from that directory, that becomes the directory in which Caddy will then look for the Caddyfile whenever it starts. 

Though, as noted [here](https://caddyserver.com/docs/caddyfile): "The Caddyfile is just a [config adapter](https://caddyserver.com/docs/config-adapters) for Caddy." So, I believe the caddy file gets converted to json and then stored in a standard location, but typically people use the Caddyfile and really it can be put wherever someone would like it to reside. I know that doesn't really help much when trying to standardize a config for something like the highlighter, etc, though.

**Comment by iliarostovtsev** [comment]
Okay, I'll come up with something usable!

What about setting code highlight for Caddy to JSON how does it look?

Could you just share the config file example?

**Comment by iliarostovtsev** [comment]
@swelljoe I think _CodeMirror_ must be removed from _Filemin_ package, and handled by a theme, because many modules have config file editors but it doesn't mean that each module individually must include it.

**Comment by MostHated** [comment]
Here is a full example of one of mine. Json was one of the first I tried, but hotkeys don't work for it, and highlighting is mostly hit or miss.

<details>
<summary>Config</summary>

Caddyfile
```shell
{
	# auto_https off
	email my@email.com # For letsencrypt certificates
}

example1.com {
	root * /home/example1/public_html

    encode gzip
    php_fastcgi unix//run/php/php7.2-fpm.sock
        
    # Prevent malicious PHP uploads from running
    @uploads {
        path_regexp path /uploads\/(.*)\.php
    }
    
    rewrite @uploads /

    @wp-admin {
        path  not ^\/wp-admin/*
    }
}

https://sub.example2.com {
	encode gzip

	header {
		# Enable cross-site filter (XSS) and tell browser to block detected attacks
		X-XSS-Protection "1; mode=block"
		# Disallow the site to be rendered within a frame (clickjacking protection)
		X-Frame-Options "DENY"
		# Prevent search engines from indexing (optional)
		X-Robots-Tag "none"
		# Server name removing
		-Server
	}

	# The negotiation endpoint is also proxied to Rocket
	reverse_proxy /notifications/hub/negotiate localhost:4577

	# Notifications redirected to the websockets server
	reverse_proxy /notifications/hub <SERVER>:3012

	reverse_proxy localhost:4577 {
		header_up Host {host}
		header_up X-Real-IP {remote}
		header_up X-Forwarded-For {remote}
		header_up X-Forwarded-Proto {scheme}
		# transport http {
		# 	tls
		# 	tls_insecure_skip_verify
		# }
	}
}
```

</details>

<details>
<summary>Highlighting Json</summary>

![](https://i.imgur.com/5ypNtX3.png)

</details>

**Comment by iliarostovtsev** [comment]
I'll update this ticket when it's done.

**Comment by iliarostovtsev** [comment]
Okay, this has been implemented on the latest commit for the theme 19.70-4.

Additionally, I finally added support to highlight files correctly without explicit mode, like Apache, Nginx, `.fcgi`, `php-fpm.conf` and etc, if opened from pop-up file editor and/or from favorites.

Give it a try.

**Note:** You would need the latest [dev](http://download.webmin.com/devel/) version of Webmin installed first, and only then update Authentic Theme to the latest using configuration page or update script.
